Adds an optional celery worker

msheiny opened the pull request:
This PR adds an optional celery worker into the mix. It won't install the actual messaging queue (like a rabbitmq) - that's up to another role/playbook to handle.

It is also up to the author of the playbook that takes advantage of this feature to add `celery` to their requirements file and add appropriate celery settings to the environment and to their django code-base.

msheiny committed:
Update CI reqs molecule, ansible, pip-tools, docker

msheiny committed:
Add scaffolding for celery worker (logs+run)

Slight consolidation of similar tasks for gunicorn as well.

msheiny committed:
Add optional celery worker task file

Includes systemd template. This doesn't actually install a messaging
queue, this will just setup a worker systemd service.

msheiny committed:
Update documentation for latest defaults

conorsch commented on tasks/setup.yml:
```diff
@@ -82,3 +95,21 @@
+    create: yes
+  with_nested:
+    - ['celery']
+    - ['/run/', '/var/log']
```
Why use `with_nested` on these tasks? `item[0]` resolves to a single string each time, so why not just hardcode it? Then you can use regular ol' `with_items` and the task becomes a lot more readable. Am I missing something?

msheiny replied:
I was trying to abstract this stuff in case we need to append to it later.. but yeah it just looks kind of silly

conorsch commented on tasks/celery.yml:
```diff
@@ -0,0 +1,24 @@
+    name: "{{ django_stack_celery_svc_name }}"
+    state: stopped
+  # The celery runner might not exist yet, lets not fail in that scenario
+  ignore_errors: true
```
Why do we need to stop it if it's running? Modifying the service template on-disk won't affect the running service until the service is restarted—which it will be via handler notification.

conorsch replied:
Instead I think the role should enforce state=running and enabled=yes on the celery service at the end of configuration. Imagine: no changes are made to the celery config, but the service was stopped manually on the server—as currently written, the role will *not* start celery under those conditions, but it absolutely should.

msheiny replied:
'doh !!! you are right this needs a conditional. 

> Modifying the service template on-disk won't affect the running service until the service is restarted

the larger problem is this though - i want to stop the existing service before i make underlying changes to the systemd unit file and reload it. This was that beta/alpha issue I was hitting. Say I have :  
* a celery worker service already running that references the alpha virtualenv and code-base
* new code is deployed making `beta` the new primary app code. the ansible logic reaches this stage, makes a systemd template change, and now the celery worker has started reference the beta base. This means that the alpha celery worker is still going in the background - which is bad.

> the role will not start celery under those conditions, but it absolutely should.

Shit you are totally right . Nice catch!! this part needs a refactor :(


The

conorsch commented on tasks/celery.yml:
```diff
@@ -0,0 +1,24 @@
+  template:
+    src: celery.j2
+    dest: "/etc/systemd/system/{{ django_stack_celery_svc_name }}.service"
+  notify: Start celery worker
```
As above, this should be `restart celery worker` IMO. The handler uses `daemon_reload` already, which is :+1:

conorsch commented on handlers/main.yml:
```diff
@@ -17,6 +17,13 @@
+    daemon_reload: yes
+    name: "{{ django_stack_celery_svc_name }}.service"
+    state: restarted
+    enabled: True
```
Handler should be called `restart celery worker`, since that's what it does. :ok_hand: on the `daemon_reload` option!

conorsch reviewed (changes requested):
A few tweaks in-line but all in all this is a solid addition. Naturally I have concerns about overloading this role heavily, but we're in a transition phase and this is a solid next step for the nonce.

msheiny committed:
Improve logic for stopping existing celery service

Previously logic was problematic and would always stop the celery
service regardless of changes pushed.

msheiny committed:
Address PR nits - handler names and setup logic

msheiny commented:
hey @conorsch made some changes. I think i solved your issue here but let me know if you require more tweaking. Really great catches though 👍

msheiny committed:
Disable ansible lint on celery stop task

Ansible-lint is telling me to use a handler here, no thanks buddy!

conorsch commented on tasks/celery.yml:
```diff
@@ -0,0 +1,41 @@
+      CELERYD_CHDIR="{{ active_deploy_dir }}"
+      CELERY_BIN="{{ django_stack_venv_dir }}/bin/celery"
+      CELERY_APP="{{ django_stack_app_name }}"
+    dest: /etc/default/celeryd
```
This task should notify a restart/daemon-reload handler, no?

msheiny replied:
Sooo... look below this ... we could add all that logic as a handler if you think it makes more sense. I can see it both ways. The important part is that the systemd daemon SHOULD NOT be daemon-reloaded + restarted before the current process is stopped.

msheiny replied:
oh i see what you are saying though... maybe i still need to add a check for this changing regardless

conorsch replied:
It's complicated: I think the right approach is _yes_ handler but _no_ to register-for-immediate stop.

conorsch replied:
After discussion out of band, added to both. However, the immediate-stop logic isn't a simple append operation, chaining the AND logic: instead, there's a new OR condition that's the registered result of changes to the celery worker config file.

conorsch committed:
Restarts celery worker on config changes

The celery worker has its own config file, separate from the systemd
service file, and changes to that should trigger service bounces.
The logic is a bit convoluted here due to the a/b service naming, but in
short we want to halt the celery worker BEFORE handlers run to make sure
the correct services are referenced in all places.

conorsch commented:
Pushed up one final commit to smooth out the handler story. Waiting on CI now, then merging.
